File: frontend_compose_old/src/jsMain/kotlin/AppConfig.kt
import core.DI
import core.InMemoryLogger
import core.Initializable
import features.text_chat.data.MessageRepository

// flavor и Flavors генерируются в build.gradle.kts
// internal enum class Flavors { Debug, Prod }
// internal val flavor = Flavors.$flavor

/**
 * Инжекция и асинхронная инициализация глобальных зависимостей.
 * Запускается на стартовой странице, ошибки выводятся туда же.
 */
suspend fun initApp() {
    DI.putAndInit(InMemoryLogger(storeUpToEntries = 100))
    DI.putAndInit(if (flavor == Flavors.Debug) DebugAppConfig() else ProdAppConfig())
    DI.putAndInit(UserSession())
    DI.put(MessageRepository())
}

abstract class AppConfig : Initializable {
    abstract val apiUrl: String
    val textChatUrl = "messages"
    
    // Информация о приложении
    val appName = "Local WiFi Chat"
    val flavorName: String
        get() = when (flavor) {
            Flavors.Debug -> "Debug"
            Flavors.Prod -> "Production"
        }

    // Опции для загрузки списков
    open val fetchBatchSize = 10
    open val fetchRefreshPeriodMs = 3000L

    // Ключи в LocalStorage
    val userHashStorageKey = "userHash"
    val userNameStorageKey = "userName"

    // Заголовки HTTP
    val userHashHeaderKey = "X-User-Hash"

    // Если нужна асинхронная инициализация конфига
    override suspend fun init() {}
}

class DebugAppConfig : AppConfig() {
    override val apiUrl = "http://localhost:9090/api"
    override val fetchBatchSize = 5

    override suspend fun init() {
        kotlinx.coroutines.delay(1000)
//        throw Exception("Testing AppConfig initialization error")
    }
}

class ProdAppConfig : AppConfig() {
    // В проде используем относительный урл, так как бандл фронтенда включен в бинарник бекенда
    override val apiUrl = "/api"
}

File: frontend_flutter/lib/app_config.dart
import 'package:local_wifi_chat_frontend/core/di.dart';
import 'package:local_wifi_chat_frontend/data/api_client.dart';
import 'package:local_wifi_chat_frontend/features/text_chat/data/devices_repository.dart';
import 'package:local_wifi_chat_frontend/features/text_chat/data/messages_repository.dart';
import 'package:local_wifi_chat_frontend/user_session.dart';

// Вариант конфигурации (flavor) берет из командной строки --dart-define=FLAVOR=RELASE
const _flavor = String.fromEnvironment('FLAVOR', defaultValue: 'DEBUG'); // RELEASE

class DebugConfig extends AppConfig {
  @override
  final apiUrl = 'http://localhost:9090/api/'; // 192.168.24.53

  @override
  // ignore: overridden_fields
  final fetchBatchSize = 5;

  @override
  Future<void> init() async {
    // throw "Тестовая ошибка инициализации приложения";
  }
}

class ReleaseConfig extends AppConfig {
  // В проде используем относительный урл, так как бандл фронтенда включен в бинарник бекенда
  @override
  final apiUrl = '/api/';
}

// Инжектит глобальные объекты, и инициализирует их ПОСЛЕДОВАТЕЛЬНО.
// Запускается в main(), ошибки выводятся туда же, логгер инициализируем раньше.
Future<void> initApp() async {
  final config = switch (_flavor) {
    'DEBUG' => DebugConfig(),
    'RELEASE' => ReleaseConfig(),
    _ => throw 'Undefined FLAVOR = $_flavor',
  };
  await DI.putAndInit(config);
  await DI.putAndInit(UserSession());
  await DI.putAndInit(
    ApiClient(
      apiUrl: config.apiUrl,
      getClientHashHeader: () => {config.userHashHeaderKey: di<UserSession>().userHash},
    ),
  );
  DI.putAll([MessagesRepository(), DevicesRepository()]);
}

/// Полиморфизм конфига - это важная и часто встречающаяся задача.
abstract class AppConfig implements Initializable {
  final appName = 'Локальный WiFi чат';
  abstract final String apiUrl;

  /// Цитируемый текст будет обрезаться до
  final maxMessageQuoteLength = 300;

  // Заголовки http
  final userHashHeaderKey = 'X-User-Hash';

  // Ключи в локальном хранилище
  final userHashKey = 'userHash';
  final userNameKey = 'userName';

  // Опции загрузки списков
  final fetchBatchSize = 10;
  final refreshListsEvery = const Duration(seconds: 3);

  String get flavor => _flavor;

  // Если нужна асинхронная инициализация (например чтение из env)
  @override
  Future<void> init() async {}
}
